Upstream interaction region code.

megangardner opened the pull request:
#### 4d00e64fc9efed19b64f488ee5b6a4e7cd5a7d47
<pre>
Upstream interaction region code.
<a href="https://bugs.webkit.org/show_bug.cgi?id=258099">https://bugs.webkit.org/show_bug.cgi?id=258099</a>
rdar://110808373

Reviewed by Aditya Keerthi.

* Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::interactionRegionLayerClass):
(WebKit::interactionRegionEffectUserInfo):
(WebKit::configureLayerForInteractionRegion):
(WebKit::configureLayerAsGuard):
(interactionRegionLayerClass): Deleted.
(configureLayerForInteractionRegion): Deleted.
(configureLayerAsGuard): Deleted.

Canonical link: <a href="https://commits.webkit.org/265214@main">https://commits.webkit.org/265214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d55e4547c6a02d9ac7bcb59e1575f339131016b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10283 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10787 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11930 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10295 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10471 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/12845 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10441 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/11177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12316 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/8485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9312 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/9583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9463 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12711 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/9915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/9074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2463 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/13324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9754 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -31,22 +31,77 @@
+SOFT_LINK_CLASS_OPTIONAL(RealitySystemSupport, RCPGlowEffectLayer)
+SOFT_LINK_CONSTANT_MAY_FAIL(RealitySystemSupport, RCPAllowedInputTypesUserInfoKey, const NSString *)
+
+@interface CALayer ()
+@property (nonatomic) CGFloat sizeMultiplier;
+@property (nonatomic, copy) void (^effectGroupConfigurator)(CARemoteEffectGroup *group);
+@end
```
This shouldn't be necessary, since you're importing `QuartzCoreSPI.h`?

megangardner replied:
Shouldn't be, but it wouldn't compile without it. Hoping to remove it when I remove the softlink as well.

pxlcoder replied:
Ah, it's in an `RealitySystemSupport` header. Though I remain surprised that the `RealitySystemSupport/RealitySystemSupport.h` doesn't bring it in.

I would just remove the declaration in `QuartzCoreSPI.h` and then we can put things in the right place when fixing the soft linking.

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -31,22 +31,77 @@
-static void configureLayerForInteractionRegion(CALayer *, NSString *) { }
-static void configureLayerAsGuard(CALayer *, NSString *) { }
-#endif
+//// FIXME: rdar://105775731
+SOFT_LINK_PRIVATE_FRAMEWORK_OPTIONAL(RealitySystemSupport)
+SOFT_LINK_CLASS_OPTIONAL(RealitySystemSupport, RCPGlowEffectLayer)
+SOFT_LINK_CONSTANT_MAY_FAIL(RealitySystemSupport, RCPAllowedInputTypesUserInfoKey, const NSString *)
```
I know you also mentioned this offline, but just leaving a note here during review, that we should remove this soft link at some point.

megangardner replied:
agreed

pxlcoder commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -31,22 +31,77 @@
 NSString *interactionRegionTypeKey = @"WKInteractionRegionType";
 NSString *interactionRegionGroupNameKey = @"WKInteractionRegionGroupName";
 
+#if PLATFORM(VISION)
```
Do we need this conditional? Since `ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)` is equivalent to `PLATFORM(VISION)`.

Noticed that you didn't include any guards around `#import <RealitySystemSupport/RealitySystemSupport.h>`, so we should probably remove to be consistent (or also add a guard there).

pxlcoder commented on Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
```diff
@@ -153,6 +157,10 @@ typedef struct _CARenderContext CARenderContext;
 @property (getter=isSeparated) BOOL separated;
 #endif
 @property BOOL toneMapToStandardDynamicRange;
+#if PLATFORM(VISION)
+@property (nonatomic) CGFloat sizeMultiplier;
+@property (nonatomic, copy) void (^effectGroupConfigurator)(CARemoteEffectGroup *group);
+#endif
```
We don't need this since it's being declared in the implementation file, and doesn't belong here anyway (these are in a RealitySystemSupport header).

pxlcoder commented on Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
```diff
@@ -57,6 +57,10 @@
 #endif
 
+#if PLATFORM(VISION)
+#import <QuartzCore/CARemoteEffectPrivate.h>
```
Is there anything we're using from this private header that can/should be declared in the open source section below? The things you added below don't really belong here, since they are not in this header or QuartzCore.

megangardner committed:
Upstream interaction region code.
https://bugs.webkit.org/show_bug.cgi?id=258099
rdar://110808373

Reviewed by Aditya Keerthi.

* Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::interactionRegionLayerClass):
(WebKit::interactionRegionEffectUserInfo):
(WebKit::configureLayerForInteractionRegion):
(WebKit::configureLayerAsGuard):
(interactionRegionLayerClass): Deleted.
(configureLayerForInteractionRegion): Deleted.
(configureLayerAsGuard): Deleted.

Canonical link: https://commits.webkit.org/265214@main
